rust/cargo-psibase/src/link.rs

```diff
@@ -1,463 +1,704 @@
+        };
+
+        // Check function types match
+        if source.func_type(source_func.id()) != dest.func_type(dest_func.id()) {
+            error = Some(anyhow!("Destination import {} matches source export by name but not by type.", name));
+        }
+
+        Some(PolyfillTarget{
+            source_fid: source_func.id(),
+            dest_fid: dest_func.id(),
+        })
+    }).collect();
+
+    if error.is_some() {
+        return Err(error.unwrap());
+    }
+
+    return Ok(Some(targets));
+}
+
+/// Gets the FunctionID if it exists for a function in a `wasm_module` based on a 
+/// `function_module` and `name`.
+fn get_import_fid(function_module : &str, name : &str, wasm_module : &Module) -> Result<Option<FunctionId>, anyhow::Error> {
+    wasm_module
+        .imports
         .iter()
-        .filter_map(|entry| {
-            if let External::Function(ty) = entry.external() {
-                Some(ImportFunction {
-                    module: entry.module().to_owned(),
-                    field: entry.field().to_owned(),
-                    ty: types[*ty as usize].clone(),
-                })
+        .find(|&i| &i.name == name && &i.module == function_module)
+        .map_or(Ok(None), |i| match i.kind {
+            walrus::ImportKind::Function(fid) => Ok(Some(fid)),
+            _ => Err(anyhow!("Import kind mismatch")),
+        })
+}
+
+/// Get the corresponding FunctionID in the `dest` module for a given FunctionID from the `source`
+/// module. If the function does not exist in the `dest` module, it will be added.
+fn to_dest_func(source_fid : FunctionId, source : &Module, dest : &mut Module ) -> Result<FunctionId, anyhow::Error> {
+
+    let f = source.funcs.get(source_fid);
+    let (source_func_kind, source_func_ty) = (&f.kind, f.ty());
+
+    let new_fid = match &source_func_kind {
+        Import(_) => {
+            let s = source.imports.get_imported_func(source_fid).unwrap();
+            let (module, name) = (&s.module, &s.name);
+
+            let dest_fid_opt = get_import_fid(module, name, &dest)?;
+
+            if let Some(dest_fid) = dest_fid_opt {
+                // The import already exists in the dest
+                if dest.func_type(dest_fid) != source.func_type(source_fid) {
+                    return Err(anyhow!("Type mismatch between imports"));
+                }
+                dest_fid
             } else {
-                None
+                // The import needs to be added to the dest. Update the called fid.
+                let ty = source.types.get(source_func_ty);
+                let (params, results) = (ty.params().to_vec(), ty.results().to_vec());
+                let dest_type = dest.types.add(&params, &results);
+                dest.add_import_func(module, name, dest_type).0
             }
-        })
-        .collect()
+        },
+        Local(_) => {
+            let new_local_func = copy_func(source_fid, source, dest)?;
+            dest.funcs.add_local(new_local_func)
```

Comment:
I don't see anything protecting against duplicating functions from the polyfill.

Reply:
Fixed